File: source/ts/src/calendar/yearly.ts
import { getDailyLabels } from '../core/panchang';
import {
  getUpcomingSolarEclipse, getUpcomingLunarEclipse,
} from '../astronomy/eclipse';
import { computeSunrise, computeSunset } from '../astronomy/sunrise';
import { getSiderealSunLongitude } from '../astronomy/sun';
import { getSiderealMoonLongitude } from '../astronomy/moon';
import { getTithiIndexFromLons } from '../core/tithi';
import { PanchangError } from '../types/errors';
import {
  resolveUtcOffset, utcToLocalDisplay, utcDateMs, wallClockToUtc, localYearWindow,
  civilDayStepper, clampToSupported, civilDayValue,
} from '../utils/timezone';
import { validateLocation, validateDate, validateLocalYearWindow } from '../utils/validation';
import type { GeoLocation } from '../types/location';
import type { AyanamsaType, MasaSystem, FestivalRegion, LegacyFestivalRegion } from '../types/options';
import type { EclipseInfo, FestivalInfo } from '../types/elements';
import { resolveMasaName } from '../i18n/resolver';

const DAY_MS = 24 * 3600_000;

export interface YearlyListingOptions {
  /** UTC offset in minutes (e.g. 330 for IST), or an IANA zone name. */
  timezone: number | string;
  ayanamsa?: AyanamsaType;
  masaSystem?: MasaSystem;
  language?: 'en' | 'hi';
  region?: FestivalRegion | LegacyFestivalRegion;
}

export interface FestivalDay {
  /**
   * The instant the day was queried at: its local midnight from `computeFestivalsForYear` (never
   * before 1900-01-01T00:00Z), the range start's local time of day from `computeFestivalsInRange`.
   * Read its date in the listing's timezone, not with `toISOString()`.
   */
  date: Date;
  festival: FestivalInfo;
}

export interface SankrantiEvent {
  /**
   * Civil day the transit is observed on, per the almanac's sunset rule, as the UTC midnight that falls
   * within that local day: the date's own UTC midnight at or east of UTC, the next one west of it, so
   * read it in the listing's timezone.
   */
  date: Date;
  moment: Date;
  /** Rashi the Sun transitioned INTO (0..11). */
  rashi: number;
  rashiName: string;
}

/**
 * Smarta Ekadashi fast dates in a Gregorian year (the local calendar year in `options.timezone`); no
 * arunodaya (Dashami-viddha) deferral. Each is the UTC midnight that falls within the local day of
 * the fast, so read it in that zone: the UTC midnight of the fast's own date at every offset at or
 * east of UTC, and the next UTC midnight west of it.
 */
export function computeEkadashiDatesForYear(
  year: number,
  location: GeoLocation,
  options: YearlyListingOptions,
): Date[] {
  validateLocation(location);
  if (!Number.isInteger(year)) throw new RangeError(`year must be integer, got ${year}`);
  const out: Date[] = [];
  const dayMs = 24 * 3600_000;
  const first = utcDateMs(year, 0, 1);
  const last = utcDateMs(year, 11, 31);
  const ayanamsa = options.ayanamsa ?? 'lahiri';

  const days: Array<{ d: number; tithi: number }> = [];
  let offset: number | undefined;
  for (let d = first - dayMs; d <= last + 2 * dayMs; d += dayMs) {
    const [midnight, dayOffset] = wallClockToUtc(d, options.timezone, offset);
    offset = dayOffset;
    if (Math.floor((midnight + dayOffset * 60_000) / dayMs) * dayMs !== d) continue;
    let sunriseUtc: Date;
    try {
      sunriseUtc = computeSunrise(new Date(midnight), location);
      computeSunrise(computeSunset(sunriseUtc, location), location);
    } catch (e: unknown) {
      if (e instanceof PanchangError && (e.code === 'NO_SUNRISE' || e.code === 'NO_SUNSET')) continue;
      throw e;
    }
    const tithiIndex = getTithiIndexFromLons(
      getSiderealMoonLongitude(sunriseUtc, ayanamsa),
      getSiderealSunLongitude(sunriseUtc, ayanamsa),
    );
    days.push({ d, tithi: tithiIndex });
  }

  for (let i = 0; i < days.length; i++) {
    const { d, tithi } = days[i]!;
    const nextEntry = days[i + 1];
    const next = nextEntry !== undefined && nextEntry.d - d === dayMs
      ? nextEntry.tithi
      : undefined;
    let fast: number | undefined;
    if ((tithi === 10 || tithi === 25) && next !== tithi) {
      fast = d;
      if ((next === 12 || next === 27) && i > 0 && d - days[i - 1]!.d === dayMs) {
        fast = days[i - 1]!.d;
      }
    } else if ((tithi === 9 && next === 11) || (tithi === 24 && next === 26)) {
      fast = d;
    }
    if (fast !== undefined && fast >= first && fast <= last) {
      out.push(new Date(civilDayValue(fast, options.timezone)));
    }
  }
  return out;
}

/** All 12 Sankranti (solar transit) dates for a Gregorian year, one per rashi. */
export function computeSankrantisForYear(
  year: number,
  location: GeoLocation,
  options: YearlyListingOptions,
): SankrantiEvent[] {
  validateLocation(location);
  if (!Number.isInteger(year)) throw new RangeError(`year must be integer, got ${year}`);
  const lang = options.language ?? 'en';
  const ayanamsa = options.ayanamsa ?? 'lahiri';
  const dayMs = 24 * 3600_000;
  const offsetAt = (ms: number) => resolveUtcOffset(options.timezone, new Date(ms));
  const sunAt = (ms: number) => getSiderealSunLongitude(new Date(ms), ayanamsa);
  const rashiAt = (ms: number) => Math.floor(sunAt(ms) / 30) % 12;

  // The scan grid's offset only aligns the daily probes: each transit's year comes from its own day.
  const gridOffset = offsetAt(utcDateMs(year, 6, 1));
  const scanStart = utcDateMs(year, 0, 1) - gridOffset * 60_000 - dayMs;
  const scanEnd = utcDateMs(year, 11, 31) + (23 * 60 + 59) * 60_000 - gridOffset * 60_000 + dayMs;

  const out: SankrantiEvent[] = [];
  let prevMs = scanStart;
  let prevRashi = rashiAt(prevMs);
  for (let t = scanStart + dayMs; t <= scanEnd; t += dayMs) {
    const rashi = rashiAt(t);
    if (rashi === prevRashi) { prevMs = t; prevRashi = rashi; continue; }

    let lo = prevMs, hi = t;
    while (hi - lo > 1000) {
      const mid = Math.floor((lo + hi) / 2);
      if (rashiAt(mid) === prevRashi) lo = mid; else hi = mid;
    }
    const transitUtc = new Date(hi);

    let anchor: Date;
    try {
      let dayStart = computeSunrise(new Date(hi - 30 * 3600_000), location);
      for (let i = 0; i < 3; i++) {
        const next = computeSunrise(computeSunset(dayStart, location), location);
        if (next.getTime() <= hi) dayStart = next; else break;
      }
      const dayEnd = computeSunset(dayStart, location);
      anchor = hi <= dayEnd.getTime()
        ? dayStart
        : computeSunrise(dayEnd, location);
    } catch (e: unknown) {
      if (!(e instanceof PanchangError && (e.code === 'NO_SUNRISE' || e.code === 'NO_SUNSET'))) throw e;
      anchor = transitUtc;
    }

    const local = utcToLocalDisplay(anchor, offsetAt(anchor.getTime()));
    const date = new Date(civilDayValue(utcDateMs(
      local.getUTCFullYear(), local.getUTCMonth(), local.getUTCDate(),
    ), options.timezone));
    if (local.getUTCFullYear() === year) {
      out.push({ date, moment: transitUtc, rashi, rashiName: resolveMasaName(rashi, lang) });
    }

    prevMs = t;
    prevRashi = rashi;
  }
  return out;
}

/** Every festival emission in an inclusive range, one civil day of `options.timezone` at a time. */
export function computeFestivalsInRange(
  start: Date,
  end: Date,
  location: GeoLocation,
  options: YearlyListingOptions,
): FestivalDay[] {
  validateDate(start);
  validateDate(end);
  validateLocation(location);
  if (start.getTime() > end.getTime()) {
    throw new RangeError(`start (${start.toISOString()}) must be ≤ end (${end.toISOString()})`);
  }
  return festivalsOnCivilDays(start.getTime(), end.getTime(), location, options);
}

function festivalsOnCivilDays(
  startMs: number,
  endMs: number,
  location: GeoLocation,
  options: YearlyListingOptions,
): FestivalDay[] {
  const out: FestivalDay[] = [];
  const dayOptions = {
    ...options,
    sections: ['festivals', 'eclipse'] as const,
    computeEndTimes: false,
  };
  const next = civilDayStepper(startMs, options.timezone);
  for (let t = next(); t <= endMs; t = next()) {
    const day = getDailyLabels(new Date(clampToSupported(t)), location, dayOptions);
    if (day === null) continue;
    for (const f of day.festivals) {
      out.push({ date: day.date, festival: f });
    }
  }
  return out;
}

/**
 * The next `count` eclipses after `fromDate`, solar and lunar merged by peak. Only `fromDate` is checked against
 * 1900..2100: the walk is not clipped at 2100 and runs until it has `count` eclipses (the Go port has no default
 * count and adds a cancellable `GetUpcomingEclipsesContext`).
 */
export function getUpcomingEclipses(
  fromDate: Date,
  location: GeoLocation,
  count: number = 5,
): EclipseInfo[] {
  validateDate(fromDate);
  validateLocation(location);
  if (!Number.isInteger(count) || count < 1) {
    throw new RangeError(`count must be a positive integer, got ${count}`);
  }
  const yearsAhead = Math.max(3, Math.ceil(count));
  const withinDays = yearsAhead * 366;

  const collected: EclipseInfo[] = [];
  let cursor = new Date(fromDate.getTime());

  while (collected.length < count) {
    const sol = getUpcomingSolarEclipse(cursor, location, withinDays);
    const lun = getUpcomingLunarEclipse(cursor, location, withinDays);
    if (!sol && !lun) break;
    let next: EclipseInfo;
    if (!sol) next = lun!;
    else if (!lun) next = sol;
    else next = sol.peak.getTime() < lun.peak.getTime() ? sol : lun;

    collected.push(next);
    cursor = new Date(next.end.getTime() + 1000);
  }

  return collected;
}

/** Every eclipse peaking in a range, sorted by `peak`; solar ones are location-filtered. */
export function computeEclipsesInRange(
  start: Date,
  end: Date,
  location: GeoLocation,
): EclipseInfo[] {
  validateDate(start);
  validateDate(end);
  validateLocation(location);
  if (start.getTime() > end.getTime()) {
    throw new RangeError(
      `start (${start.toISOString()}) must be ≤ end (${end.toISOString()})`,
    );
  }

  const spanDays = Math.ceil((end.getTime() - start.getTime()) / DAY_MS) + 1;
  const startMs = start.getTime();
  const endMs = end.getTime();
  const maxSteps = Math.ceil(spanDays / 20) + 50;

  // A walk only finds syzygies at or after `start`; the search from a day earlier adds one that peaks after it.
  const walk = (next: (from: Date, withinDays: number) => EclipseInfo | null): EclipseInfo[] => {
    const acc: EclipseInfo[] = [];
    let cursor = new Date(startMs);
    for (let step = 0; step < maxSteps; step++) {
      const e = next(cursor, spanDays);
      if (!e || e.peak.getTime() > endMs) break;
      if (e.peak.getTime() >= startMs) acc.push(e);
      cursor = new Date(e.end.getTime() + 1000);
    }
    const early = next(new Date(startMs - DAY_MS), 2);
    if (
      early && early.peak.getTime() >= startMs && early.peak.getTime() <= endMs &&
      !acc.some(e => Math.abs(e.peak.getTime() - early.peak.getTime()) < DAY_MS)
    ) {
      acc.unshift(early);
    }
    return acc;
  };

  const solar = walk((from, days) => getUpcomingSolarEclipse(from, location, days));
  const lunar = walk((from, days) => getUpcomingLunarEclipse(from, location, days));

  return [...solar, ...lunar].sort((a, b) => a.peak.getTime() - b.peak.getTime());
}

/** Every festival emission in local calendar year `year`, each day queried at its local midnight. */
export function computeFestivalsForYear(
  year: number,
  location: GeoLocation,
  options: YearlyListingOptions,
): FestivalDay[] {
  if (!Number.isInteger(year)) throw new RangeError(`year must be integer, got ${year}`);
  const [start, end] = localYearWindow(year, options.timezone);
  validateLocalYearWindow(year, start, end);
  validateLocation(location);
  return festivalsOnCivilDays(start, end, location, options);
}

/** Every eclipse peaking in local calendar year `year`, as {@link computeEclipsesInRange} lists them. */
export function computeEclipsesForYear(
  year: number,
  location: GeoLocation,
  options: { timezone: number | string },
): EclipseInfo[] {
  if (!Number.isInteger(year)) throw new RangeError(`year must be integer, got ${year}`);
  const [start, end] = localYearWindow(year, options.timezone);
  validateLocalYearWindow(year, start, end);
  return computeEclipsesInRange(
    new Date(clampToSupported(start)), new Date(clampToSupported(end)), location,
  );
}

/** @deprecated Renamed to {@link computeEkadashiDatesForYear} in v5. */
export const getEkadashiDatesForYear = computeEkadashiDatesForYear;

/** @deprecated Renamed to {@link computeSankrantisForYear} in v5. */
export const getSankrantisForYear = computeSankrantisForYear;

/** @deprecated Renamed to {@link computeFestivalsInRange} in v5. */
export const getFestivalsInRange = computeFestivalsInRange;

/** @deprecated Renamed to {@link computeEclipsesInRange} in v5. */
export const getEclipsesInRange = computeEclipsesInRange;

File: source/go/internal/calendar/yearly.go
package calendar

import (
	"context"
	"errors"
	"math"
	"sort"

	"github.com/ishankgupta95/panchang/source/go/v5/internal/astronomy"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/core"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/i18n"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/utils"
	"github.com/ishankgupta95/panchang/source/go/v5/types"
)

const dayMs int64 = 24 * 3600_000

type ekadashiDay struct {
	dMs   int64
	tithi int
}

func ComputeEkadashiDatesForYear(ctx context.Context,
	eph *astronomy.EphemerisCtx,
	year int,
	location types.GeoLocation,
	options YearlyListingOptions,
) ([]types.JSDate, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	if err := utils.ValidateLocation(location); err != nil {
		return nil, err
	}
	out := []types.JSDate{}
	first := utils.UtcDateMs(year, 0, 1)
	last := utils.UtcDateMs(year, 11, 31)
	ayanamsa := yearlyResolvedAyanamsa(options)

	days := make([]ekadashiDay, 0, 370)
	var offset int
	var hint *int
	for d := first - dayMs; d <= last+2*dayMs; d += dayMs {
		if err := ctx.Err(); err != nil {
			return nil, err
		}
		midnight, dayOffset, err := utils.WallClockToUtc(d, options.Timezone, hint)
		if err != nil {
			return nil, err
		}
		offset, hint = dayOffset, &offset
		if floorDiv(midnight+int64(dayOffset)*60_000, dayMs)*dayMs != d {
			continue
		}
		sunriseUtcMs, err := astronomy.ComputeSunrise(eph, midnight, location,
			astronomy.DefaultRiseSetLimitDays)
		if err == nil {
			var sunsetUtcMs int64
			sunsetUtcMs, err = astronomy.ComputeSunset(eph, sunriseUtcMs, location,
				astronomy.DefaultRiseSetLimitDays)
			if err == nil {
				_, err = astronomy.ComputeSunrise(eph, sunsetUtcMs, location,
					astronomy.DefaultRiseSetLimitDays)
			}
		}
		if err != nil {
			if isPolarRiseSetError(err) {
				continue
			}
			return nil, err
		}

		siderealMoon, err := astronomy.GetSiderealMoonLongitude(eph, sunriseUtcMs, ayanamsa)
		if err != nil {
			return nil, err
		}
		siderealSun, err := astronomy.GetSiderealSunLongitude(eph, sunriseUtcMs, ayanamsa)
		if err != nil {
			return nil, err
		}
		days = append(days, ekadashiDay{dMs: d, tithi: core.GetTithiIndexFromLons(siderealMoon, siderealSun)})
	}

	for i := 0; i < len(days); i++ {
		d, tithi := days[i].dMs, days[i].tithi
		next, hasNext := 0, false
		if i+1 < len(days) && days[i+1].dMs-d == dayMs {
			next, hasNext = days[i+1].tithi, true
		}
		fast, isFast := int64(0), false
		switch {
		case (tithi == 10 || tithi == 25) && !(hasNext && next == tithi):
			fast, isFast = d, true
			if hasNext && (next == 12 || next == 27) && i > 0 && d-days[i-1].dMs == dayMs {
				fast = days[i-1].dMs
			}
		case hasNext && ((tithi == 9 && next == 11) || (tithi == 24 && next == 26)):
			fast, isFast = d, true
		}
		if isFast && fast >= first && fast <= last {
			value, err := utils.CivilDayValue(fast, options.Timezone)
			if err != nil {
				return nil, err
			}
			out = append(out, types.Date(value))
		}
	}
	return out, nil
}

func floorDiv(a, b int64) int64 {
	q := a / b
	if a%b != 0 && (a < 0) != (b < 0) {
		q--
	}
	return q
}

// isPolarRiseSetError matches by code, not against the exported Sentinel
// variables, whose Code any importer can overwrite.
func isPolarRiseSetError(err error) bool {
	var pe *types.PanchangError
	return errors.As(err, &pe) && (pe.Code == types.ErrNoSunrise || pe.Code == types.ErrNoSunset)
}

func ComputeSankrantisForYear(ctx context.Context,
	eph *astronomy.EphemerisCtx,
	year int,
	location types.GeoLocation,
	options YearlyListingOptions,
) ([]SankrantiEvent, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	if err := utils.ValidateLocation(location); err != nil {
		return nil, err
	}
	lang := yearlyResolvedLanguage(options)
	ayanamsa := yearlyResolvedAyanamsa(options)
	// The scan grid's offset only aligns the daily probes: each transit's year comes from its own day.
	gridOffset, err := utils.ResolveUtcOffset(options.Timezone, utils.UtcDateMs(year, 6, 1))
	if err != nil {
		return nil, err
	}
	rashiAt := func(ms int64) (int, error) {
		lon, err := astronomy.GetSiderealSunLongitude(eph, ms, ayanamsa)
		if err != nil {
			return 0, err
		}
		return int(math.Floor(lon/30)) % 12, nil
	}

	scanStart := utils.UtcDateMs(year, 0, 1) - int64(gridOffset)*60_000 - dayMs
	scanEnd := utils.UtcDateMs(year, 11, 31) + 23*3600_000 + 59*60_000 -
		int64(gridOffset)*60_000 + dayMs

	out := []SankrantiEvent{}
	prevMs := scanStart
	prevRashi, err := rashiAt(prevMs)
	if err != nil {
		return nil, err
	}
	for t := scanStart + dayMs; t <= scanEnd; t += dayMs {
		if err := ctx.Err(); err != nil {
			return nil, err
		}
		rashi, err := rashiAt(t)
		if err != nil {
			return nil, err
		}
		if rashi == prevRashi {
			prevMs, prevRashi = t, rashi
			continue
		}

		lo, hi := prevMs, t
		for hi-lo > 1000 {
			mid := (lo + hi) >> 1
			r, err := rashiAt(mid)
			if err != nil {
				return nil, err
			}
			if r == prevRashi {
				lo = mid
			} else {
				hi = mid
			}
		}
		transitUtcMs := hi

		anchorMs, anchorErr := sankrantiAnchor(eph, hi, location)
		if anchorErr != nil {
			if !isPolarRiseSetError(anchorErr) {
				return nil, anchorErr
			}
			anchorMs = transitUtcMs
		}

		anchorOffset, err := utils.ResolveUtcOffset(options.Timezone, anchorMs)
		if err != nil {
			return nil, err
		}
		local := types.Date(utils.UtcToLocalDisplay(anchorMs, anchorOffset))
		date, err := utils.CivilDayValue(
			utils.UtcDateMs(local.UTCFullYear(), local.UTCMonth(), local.UTCDate()), options.Timezone)
		if err != nil {
			return nil, err
		}
		if local.UTCFullYear() == year {
			out = append(out, SankrantiEvent{
				Date:      types.Date(date),
				Moment:    types.Date(transitUtcMs),
				Rashi:     rashi,
				RashiName: i18n.ResolveMasaName(rashi, lang),
			})
		}

		prevMs, prevRashi = t, rashi
	}
	return out, nil
}

func sankrantiAnchor(
	eph *astronomy.EphemerisCtx, transitMs int64, location types.GeoLocation,
) (int64, error) {
	dayStart, err := astronomy.ComputeSunrise(eph, transitMs-30*3600_000, location,
		astronomy.DefaultRiseSetLimitDays)
	if err != nil {
		return 0, err
	}
	for i := 0; i < 3; i++ {
		sunset, err := astronomy.ComputeSunset(eph, dayStart, location,
			astronomy.DefaultRiseSetLimitDays)
		if err != nil {
			return 0, err
		}
		next, err := astronomy.ComputeSunrise(eph, sunset, location,
			astronomy.DefaultRiseSetLimitDays)
		if err != nil {
			return 0, err
		}
		if next <= transitMs {
			dayStart = next
		} else {
			break
		}
	}
	dayEnd, err := astronomy.ComputeSunset(eph, dayStart, location,
		astronomy.DefaultRiseSetLimitDays)
	if err != nil {
		return 0, err
	}
	if transitMs <= dayEnd {
		return dayStart, nil
	}
	return astronomy.ComputeSunrise(eph, dayEnd, location, astronomy.DefaultRiseSetLimitDays)
}

func ComputeFestivalsInRange(ctx context.Context,
	eph *astronomy.EphemerisCtx,
	startMs, endMs int64,
	location types.GeoLocation,
	options YearlyListingOptions,
) ([]FestivalDay, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	if err := utils.ValidateDate(startMs); err != nil {
		return nil, err
	}
	if err := utils.ValidateDate(endMs); err != nil {
		return nil, err
	}
	if err := utils.ValidateLocation(location); err != nil {
		return nil, err
	}
	if startMs > endMs {
		return nil, types.Codef(types.ErrInvalidInput,
			"start (%s) must be ≤ end (%s)",
			types.Date(startMs).ISOString(), types.Date(endMs).ISOString())
	}
	return festivalsOnCivilDays(ctx, eph, startMs, endMs, location, options)
}

func festivalsOnCivilDays(ctx context.Context,
	eph *astronomy.EphemerisCtx,
	startMs, endMs int64,
	location types.GeoLocation,
	options YearlyListingOptions,
) ([]FestivalDay, error) {
	out := []FestivalDay{}
	opts := yearlyPanchangOptions(options)
	opts.Sections = core.Sections(core.SectionFestivals, core.SectionEclipse)
	opts.SectionsGiven = true
	computeEndTimes := false
	opts.ComputeEndTimes = &computeEndTimes

	next, err := utils.CivilDayStepper(startMs, options.Timezone)
	if err != nil {
		return nil, err
	}
	for {
		if err := ctx.Err(); err != nil {
			return nil, err
		}
		t, err := next()
		if err != nil {
			return nil, err
		}
		if t > endMs {
			break
		}
		day, ok, err := core.GetDailyLabels(eph, utils.ClampToSupported(t), location, opts)
		if err != nil {
			return nil, err
		}
		if !ok {
			continue
		}
		for _, f := range day.Festivals {
			out = append(out, FestivalDay{Date: day.Date, Festival: f})
		}
	}
	return out, nil
}

func GetUpcomingEclipses(
	eph *astronomy.EphemerisCtx,
	fromMs int64,
	location types.GeoLocation,
	count int,
) ([]astronomy.EclipseInfo, error) {
	return GetUpcomingEclipsesContext(context.Background(), eph, fromMs, location, count)
}

// GetUpcomingEclipsesContext is GetUpcomingEclipses checking ctx before it
// validates and before each eclipse it looks for.
func GetUpcomingEclipsesContext(ctx context.Context,
	eph *astronomy.EphemerisCtx,
	fromMs int64,
	location types.GeoLocation,
	count int,
) ([]astronomy.EclipseInfo, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	if err := utils.ValidateDate(fromMs); err != nil {
		return nil, err
	}
	if err := utils.ValidateLocation(location); err != nil {
		return nil, err
	}
	if count < 1 {
		return nil, types.Codef(types.ErrInvalidInput,
			"count must be a positive integer, got %d", count)
	}
	yearsAhead := count
	if yearsAhead < 3 {
		yearsAhead = 3
	}
	withinDays := yearsAhead * 366

	collected := []astronomy.EclipseInfo{}
	cursorMs := fromMs

	for len(collected) < count {
		if err := ctx.Err(); err != nil {
			return nil, err
		}
		sol, hasSol := astronomy.GetUpcomingSolarEclipse(eph, cursorMs, location, withinDays, types.LanguageEn)
		lun, hasLun := astronomy.GetUpcomingLunarEclipse(eph, cursorMs, location, withinDays, types.LanguageEn)
		if !hasSol && !hasLun {
			break
		}
		var next astronomy.EclipseInfo
		switch {
		case !hasSol:
			next = lun
		case !hasLun:
			next = sol
		case sol.PeakMs.Ms() < lun.PeakMs.Ms():
			next = sol
		default:
			next = lun
		}

		collected = append(collected, next)
		cursorMs = next.EndMs.Ms() + 1000
	}

	return collected, nil
}

func ComputeEclipsesInRange(ctx context.Context,
	eph *astronomy.EphemerisCtx,
	startMs, endMs int64,
	location types.GeoLocation,
) ([]astronomy.EclipseInfo, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	if err := utils.ValidateDate(startMs); err != nil {
		return nil, err
	}
	if err := utils.ValidateDate(endMs); err != nil {
		return nil, err
	}
	if err := utils.ValidateLocation(location); err != nil {
		return nil, err
	}
	if startMs > endMs {
		return nil, types.Codef(types.ErrInvalidInput,
			"start (%s) must be ≤ end (%s)",
			types.Date(startMs).ISOString(), types.Date(endMs).ISOString())
	}

	spanDays := int(math.Ceil(float64(endMs-startMs)/float64(24*3600_000))) + 1
	maxSteps := int(math.Ceil(float64(spanDays)/20)) + 50

	// A walk only finds syzygies at or after startMs; the search from a day earlier adds one that peaks after it.
	walk := func(next func(fromMs int64, withinDays int) (astronomy.EclipseInfo, bool)) ([]astronomy.EclipseInfo, error) {
		acc := []astronomy.EclipseInfo{}
		cursorMs := startMs
		for step := 0; step < maxSteps; step++ {
			if err := ctx.Err(); err != nil {
				return nil, err
			}
			e, ok := next(cursorMs, spanDays)
			if !ok || e.PeakMs.Ms() > endMs {
				break
			}
			if e.PeakMs.Ms() >= startMs {
				acc = append(acc, e)
			}
			cursorMs = e.EndMs.Ms() + 1000
		}
		if err := ctx.Err(); err != nil {
			return nil, err
		}
		early, ok := next(startMs-dayMs, 2)
		if !ok || early.PeakMs.Ms() < startMs || early.PeakMs.Ms() > endMs {
			return acc, nil
		}
		for _, e := range acc {
			if d := e.PeakMs.Ms() - early.PeakMs.Ms(); d > -dayMs && d < dayMs {
				return acc, nil
			}
		}
		return append([]astronomy.EclipseInfo{early}, acc...), nil
	}

	solar, err := walk(func(fromMs int64, withinDays int) (astronomy.EclipseInfo, bool) {
		return astronomy.GetUpcomingSolarEclipse(eph, fromMs, location, withinDays, types.LanguageEn)
	})
	if err != nil {
		return nil, err
	}
	lunar, err := walk(func(fromMs int64, withinDays int) (astronomy.EclipseInfo, bool) {
		return astronomy.GetUpcomingLunarEclipse(eph, fromMs, location, withinDays, types.LanguageEn)
	})
	if err != nil {
		return nil, err
	}

	all := make([]astronomy.EclipseInfo, 0, len(solar)+len(lunar))
	all = append(all, solar...)
	all = append(all, lunar...)
	sort.SliceStable(all, func(i, j int) bool { return all[i].PeakMs.Ms() < all[j].PeakMs.Ms() })
	return all, nil
}

func ComputeFestivalsForYear(ctx context.Context,
	eph *astronomy.EphemerisCtx,
	year int,
	location types.GeoLocation,
	options YearlyListingOptions,
) ([]FestivalDay, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	startMs, endMs, err := utils.LocalYearWindow(year, options.Timezone)
	if err != nil {
		return nil, err
	}
	if err := utils.ValidateLocalYearWindow(year, startMs, endMs); err != nil {
		return nil, err
	}
	if err := utils.ValidateLocation(location); err != nil {
		return nil, err
	}
	return festivalsOnCivilDays(ctx, eph, startMs, endMs, location, options)
}

func ComputeEclipsesForYear(ctx context.Context,
	eph *astronomy.EphemerisCtx,
	year int,
	location types.GeoLocation,
	timezone types.Timezone,
) ([]astronomy.EclipseInfo, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	startMs, endMs, err := utils.LocalYearWindow(year, timezone)
	if err != nil {
		return nil, err
	}
	if err := utils.ValidateLocalYearWindow(year, startMs, endMs); err != nil {
		return nil, err
	}
	return ComputeEclipsesInRange(ctx, eph, utils.ClampToSupported(startMs), utils.ClampToSupported(endMs), location)
}
